[Pytorch Edge] Runtime ops compatibility api

JacobSzwejbka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57570 [Pytorch Edge] Runtime ops compatibility api**

Move runtime ops compatibility api to OSS and introduce schema information

Differential Revision: [D28203104](https://our.internmc.facebook.com/intern/diff/D28203104/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28203104/)!

Commit:
[Pytorch Edge] Runtime ops compatibility api

Move runtime ops compatibility api to OSS and introduce schema information

Differential Revision: [D28203104](https://our.internmc.facebook.com/intern/diff/D28203104/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28203104/)!

[ghstack-poisoned]

JacobSzwejbka commented on torch/csrc/jit/mobile/runtime_compatibility.h:
```diff
@@ -1,11 +1,20 @@
 namespace torch {
 namespace jit {
 
+static const int NO_SCHEMA = -1;
```
this is probably not what I want to be doing. Anyone have a code pointer to how stuff like this is normally done in torch?

iseeyuan replied:
Internally I'm using c10::optional<int>. When serializing it I use -1.

JacobSzwejbka commented on torch/csrc/jit/mobile/runtime_compatibility.h:
```diff
@@ -1,11 +1,20 @@
 
+static const int NO_SCHEMA = -1;
+
+struct OperatorInfo {
```
This is the same as in 57501 I'll consolidate the two.

iseeyuan replied:
Would it be consolidated in this PR?

JacobSzwejbka replied:
I'll rebase the model api to be dependent on this one and have the model method use the runtimes definition

JacobSzwejbka commented on torch/csrc/jit/mobile/runtime_compatibility.cpp:
```diff
@@ -8,5 +10,45 @@ uint64_t _get_runtime_bytecode_version() {
+    // grab schema
+    const auto op_handle = c10::Dispatcher::singleton().findOp(op);
+    int num_schema_args = NO_SCHEMA;
+    if (op_handle->hasSchema()) {
```
@raziel @iseeyuan What do you think about not returning ops with no schema from this api and instead just collecting them and then throwing a torch warn that lists all the ops with no schema?

iseeyuan replied:
I think it may make sense to let caller know what's exactly inside the runtime. Returning all ops, with/wo schema makes more sense to me.

raziel replied:
I don't understand well the no-schema situation.

* Are these no-schema ops valid ops but that lack the metadata? i.e. can they be used to run stuff even without the schema? If so then it seems we should include them in the set.
* I think you're investigating this, but why do we have these no-schema ops? Is this something we think we can/will fix soon?

iseeyuan replied:
@raziel For your first question, yes the no-schema ops are "valid" ops that can be used. The kernel is there but the schema is stripped. For your second question, I'll do some experiment and see if there's a fix.

raziel replied:
yeah so since no-schema ops are still valid ops then we should return them in the API.

Separately we should figure out why they have no schema.

raziel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -620,6 +620,11 @@ TEST(LiteInterpreterTest, GetRuntimeByteCodeVersion) {
 
+TEST(LiteInterpreterTest, GetRuntimeOpsAndInfo) {
+  auto runtime_ops = _get_runtime_ops_and_info();
+  AT_ASSERT(runtime_ops.size() > 2900);
```
where does 2900 come from?

JacobSzwejbka replied:
I was getting the total ops in the runtime (without custom ops) to be around 3k so I just picked a number that felt unlikely for the total ops to ever go beneath. I could iterate over the ops in the map and make sure they have schemas as well, but I'm hesitant to depend on any specific op existing or its schema being a certain size.

raziel commented on torch/csrc/jit/mobile/runtime_compatibility.cpp:
```diff
@@ -8,5 +10,45 @@ uint64_t _get_runtime_bytecode_version() {
+  for (const auto& full_op : nonDispatcherOperators) {
+    auto op = full_op->schema();
+    int num_schema_args = op.arguments().size();
+    if (op.overload_name() != "") {
+      result.emplace(
+          op.name() + "." + op.overload_name(), OperatorInfo{num_schema_args});
+    } else {
+      result.emplace(op.name(), OperatorInfo{num_schema_args});
+    }
+  }
```
```suggestion
    auto op_name = op.name();
    if (op.overload_name() != "") {
      op_name += ("." + op.overload_name());
    }
    result.emplace(op_name, OperatorInfo{num_schema_args});
  }
```

raziel commented on torch/csrc/jit/mobile/runtime_compatibility.cpp:
```diff
@@ -8,5 +10,45 @@ uint64_t _get_runtime_bytecode_version() {
+      num_schema_args = op_handle->schema().arguments().size();
+    }
+
+    if (op.overload_name != "") {
+      result.emplace(
+          op.name + "." + op.overload_name, OperatorInfo{num_schema_args});
+    } else {
+      result.emplace(op.name, OperatorInfo{num_schema_args});
+    }
```
Same suggestion as above. Maybe make a little lambda function for the name?

    auto fullOpName = [](auto& op) 
    { 
        return op.overload_name() == "" ? op.name() : op.name() + "." + op.overload_name();
    };

cccclai replied:
I think it's probably more efficient to use empty function than string comparison.
```
auto fullOpName = [](auto& op) 
{ 
    return op.overload_name().empty() ? op.name() : op.name() + "." + op.overload_name();
};
```

JacobSzwejbka replied:
I dont think I can do a lambda cleanly because what type 'op' is differs between the 2 loops. In one of them I have to go through getter functions .name() and the other I have to access the variable directly .name

Commit:
Update on "[Pytorch Edge] Runtime ops compatibility api"


Move runtime ops compatibility api to OSS and introduce schema information

Differential Revision: [D28203104](https://our.internmc.facebook.com/intern/diff/D28203104/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28203104/)!

[ghstack-poisoned]

raziel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -721,6 +721,11 @@ TEST(LiteInterpreterTest, BackPortByteCodeModelAllVersions) {
 }
 
+TEST(LiteInterpreterTest, GetRuntimeOpsAndInfo) {
+  auto runtime_ops = _get_runtime_ops_and_info();
+  AT_ASSERT(runtime_ops.size() > 2900);
```
```suggestion
  auto runtime_ops = _get_runtime_ops_and_info();
  // Ballpark estimate of the minimal number of ops; just used to
  // verify API returns a reasonably large number.
  AT_ASSERT(runtime_ops.size() > 2900);
```

Commit:
Update on "[Pytorch Edge] Runtime ops compatibility api"


Move runtime ops compatibility api to OSS and introduce schema information

Differential Revision: [D28203104](https://our.internmc.facebook.com/intern/diff/D28203104/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28203104/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge] Runtime ops compatibility api"


Move runtime ops compatibility api to OSS and introduce schema information

Differential Revision: [D28203104](https://our.internmc.facebook.com/intern/diff/D28203104/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28203104/)!

[ghstack-poisoned]
